### PR TITLE
[gbinder-radio] Add support for android.hardware.radio@1.4. JB#54258

### PR DESCRIPTION
**Pull request by d-grigorev**

### CODE REVIEW

File: include/radio_util.h [hunk 1]
```diff
@@ -57,6 +57,10 @@ RADIO_RESP
 radio_req_resp(
     RADIO_REQ req);
 
+const char*
+radio_interface_name(
+    RADIO_INTERFACE interface);
```
Comment (monich) [inline include/radio_util.h]: What's the purpose of this new public API? It doesn't seem to be used by the library.

Comment (d-grigorev) [reply include/radio_util.h]: It is used by https://github.com/mer-hybris/ofono-ril-binder-plugin/pull/28.

File: include/radio_types.h [hunk 2]
```diff
@@ -994,7 +1034,7 @@ typedef struct radio_cell_info_1_4 {
         RadioCellInfoNr nr RADIO_ALIGNED(8);
     } info RADIO_ALIGNED(8);
 } RADIO_ALIGNED(8) RadioCellInfo_1_4; /* Since 1.2.5 */
-G_STATIC_ASSERT(sizeof(RadioCellInfo_1_4) == 128);
+G_STATIC_ASSERT(sizeof(RadioCellInfo_1_4) == 136);
```
Comment (monich) [inline include/radio_types.h:1023]: Good catch!

File: include/radio_types.h [hunk 3]
```diff
@@ -416,6 +431,15 @@ typedef struct radio_card_status_1_2 {
 } RADIO_ALIGNED(8) RadioCardStatus_1_2; /* Since 1.2.3 */
 G_STATIC_ASSERT(sizeof(RadioCardStatus_1_2) == 80);
 
+typedef struct radio_card_status_1_4 {
+    RadioCardStatus base RADIO_ALIGNED(8);
+    gint32 physicalSlotId RADIO_ALIGNED(4);
+    GBinderHidlString atr RADIO_ALIGNED(8);
+    GBinderHidlString iccid RADIO_ALIGNED(8);
+    GBinderHidlString eid RADIO_ALIGNED(8);
+} RADIO_ALIGNED(8) RadioCardStatus_1_4; /* Since 1.2.7 */
```
Comment (monich) [inline include/radio_types.h]: Let's make 1.4.0 the next version, to reflect radio@1.4 support.

File: include/radio_util.h (hunk 1 above)
Comment (monich) [reply include/radio_util.h]: Then mark it with `/* Since 1.4.0 */` in both .h and .c files. Let's assume that the next version is going to be 1.4.0

Comment (monich) [reply include/radio_util.h]: Actually, I don't think it's necessary. Let libgbiner-radio use the enum and the client of the library deal with string representations of the enum when and if it needs that.

Comment (d-grigorev) [reply include/radio_util.h]: Ok. I dropped last two commits including the one that moves the definition of `RADIO_INTERFACE` type.

File: include/radio_types.h (hunk 2 above)
Comment (monich) [reply include/radio_types.h:1023]: Apparently I confused `::android::hardware::radio::V1_4::CellInfo` with `::android::hardware::radio::V1_4::CellInfo::Info` 😐

File: src/radio_instance.c [hunk 4]
```diff
@@ -93,19 +93,25 @@ static GHashTable* radio_instance_table = NULL;
 #define DEFAULT_INTERFACE RADIO_INTERFACE_1_0
 
 static const GBinderClientIfaceInfo radio_iface_info[] = {
+    {RADIO_1_4, RADIO_1_4_REQ_LAST }, /* Since 1.4.0 */
+    {RADIO_1_3, RADIO_1_3_REQ_LAST },
     {RADIO_1_2, RADIO_1_2_REQ_LAST },
     {RADIO_1_1, RADIO_1_1_REQ_LAST },
     {RADIO_1_0, RADIO_1_0_REQ_LAST }
 };
```
Comment (monich) [inline src/radio_instance.c:101]: Add
```c
G_STATIC_ASSERT(G_N_ELEMENTS(radio_iface_info) == RADIO_INTERFACE_COUNT);
```
to ensure that all interfaces are covered.

File: src/radio_instance.c [hunk 5]
```diff
@@ -93,19 +93,25 @@ static GHashTable* radio_instance_table = NULL;
...
 
 static const char* const radio_indication_ifaces[] = {
+    RADIO_INDICATION_1_4,
+    RADIO_INDICATION_1_3,
     RADIO_INDICATION_1_2,
     RADIO_INDICATION_1_1,
     RADIO_INDICATION_1_0,
     NULL
 };
```
Comment (monich) [inline src/radio_instance.c:111]:
```c
G_STATIC_ASSERT(G_N_ELEMENTS(radio_indication_ifaces)==RADIO_INTERFACE_COUNT+1);
```

File: src/radio_instance.c [hunk 6]
```diff
@@ -93,19 +93,25 @@ static GHashTable* radio_instance_table = NULL;
...
 
 static const char* const radio_response_ifaces[] = {
```
Comment (monich) [inline src/radio_instance.c:114]:
```c
G_STATIC_ASSERT(G_N_ELEMENTS(radio_response_ifaces) == RADIO_INTERFACE_COUNT+1);
```
after this one.

File: src/radio_instance.c [hunk 7]
```diff
@@ -121,20 +127,32 @@ typedef struct radio_interface_desc {
...
+        radio_response_ifaces + 3
     },{
         RADIO_INTERFACE_1_0,
         RADIO_1_0,
-        radio_indication_ifaces + 2,
-        radio_response_ifaces + 2
+        radio_indication_ifaces + 4,
+        radio_response_ifaces + 4
     }
 };
```
Comment (monich) [inline src/radio_instance.c:144]:
```c
G_STATIC_ASSERT(G_N_ELEMENTS(radio_interfaces) == RADIO_INTERFACE_COUNT);
```

Comment (monich) [reply src/radio_instance.c:144]: Why it wasn't there before? When those were only a few lines of code and they all fit nicely into a single, even small, screen, it wasn't that important. Now it's getting easier to miss something.

File: include/radio_types.h (hunk 2 above)
Comment (d-grigorev) [reply include/radio_types.h:1023]: It is so easy to do.

File: src/radio_instance.c (hunk 7 above)
Comment (d-grigorev) [reply src/radio_instance.c:144]: Done.

File: src/radio_instance.c [hunk 8]
```diff
@@ -121,22 +130,35 @@ typedef struct radio_interface_desc {
```
Comment (monich) [inline src/radio_instance.c:130]: Another suggestion by @mlehtima (hopefully the last one)
```c
#define RADIO_INTERFACE_INDEX(x) (RADIO_INTERFACE_COUNT - x - 1)
```
and then hardcoded indices can be replaced with macros like this:
```c
static const RadioInterfaceDesc radio_interfaces[] = {
    {
        RADIO_INTERFACE_1_4,
        RADIO_1_4,
        radio_indication_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_4),
        radio_response_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_4)
    },{
        RADIO_INTERFACE_1_3,
        RADIO_1_3,
        radio_indication_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_3),
        radio_response_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_3)
    },{
        RADIO_INTERFACE_1_2,
        RADIO_1_2,
        radio_indication_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_2),
        radio_response_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_2)
    },{
        RADIO_INTERFACE_1_1,
        RADIO_1_1,
        radio_indication_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_1),
        radio_response_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_1)
    },{
        RADIO_INTERFACE_1_0,
        RADIO_1_0,
        radio_indication_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_0),
        radio_response_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_1_0)
    }
};
```
I would go even further and reduce it to something like this:
```c
#define RADIO_INTERFACE_DESC(v) \
        RADIO_INTERFACE_##v, RADIO_##v, \
        radio_indication_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_##v), \
        radio_response_ifaces + RADIO_INTERFACE_INDEX(RADIO_INTERFACE_##v)

static const RadioInterfaceDesc radio_interfaces[] = {
   { RADIO_INTERFACE_DESC(1_4) },
   { RADIO_INTERFACE_DESC(1_3) },
   { RADIO_INTERFACE_DESC(1_2) },
   { RADIO_INTERFACE_DESC(1_1) },
   { RADIO_INTERFACE_DESC(1_0) }
};
```

Comment (d-grigorev) [reply src/radio_instance.c:130]: After a few layers of macros, the c code starts to look like magic :)
Applied your suggestion.

**Review by monich — APPROVED** [review]
LGTM

**Comment by monich** [comment]
Thanks!

**Pull request merged**